File: packages/pigeon/platform_tests/test_plugin/example/ios/RunnerTests/MultipleArityTests.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Flutter
import Testing

@testable import test_plugin

class MockMultipleArityHostApi: MultipleArityHostApi {
  func subtract(x: Int64, y: Int64) -> Int64 {
    return x - y
  }
}

@MainActor
struct MultipleArityTests {
  var codec = FlutterStandardMessageCodec.sharedInstance()

  @Test
  func simpleHost() async throws {
    let binaryMessenger = MockBinaryMessenger<Int64>(codec: EnumPigeonCodec.shared)
    MultipleArityHostApiSetup.setUp(
      binaryMessenger: binaryMessenger, api: MockMultipleArityHostApi())
    let channelName = "dev.flutter.pigeon.pigeon_integration_tests.MultipleArityHostApi.subtract"
    #expect(binaryMessenger.handlers[channelName] != nil)

    let inputX = 10
    let inputY = 7
    let inputEncoded = binaryMessenger.codec.encode([inputX, inputY])

    await confirmation { confirmed in
      binaryMessenger.handlers[channelName]?(inputEncoded) { data in
        let outputList = binaryMessenger.codec.decode(data) as? [Any]
        #expect(outputList != nil)

        let output = outputList![0] as? Int64
        #expect(output == 3)
        #expect(outputList?.count == 1)
        confirmed()
      }
    }
  }

  @Test
  func simpleFlutter() async throws {
    let binaryMessenger = HandlerBinaryMessenger(codec: codec) { args in
      return (args[0] as! Int) - (args[1] as! Int)
    }
    let api = MultipleArityFlutterApi(binaryMessenger: binaryMessenger)

    await confirmation { confirmed in
      api.subtract(x: 30, y: 10) { result in
        switch result {
        case .success(let res):
          #expect(res == 20)
          confirmed()
        case .failure(_):
          Issue.record("Failed")
          return
        }
      }
    }
  }
}

File: packages/pigeon/platform_tests/test_plugin/android/src/test/kotlin/com/example/test_plugin/MultipleArityTests.kt
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.test_plugin

import io.flutter.plugin.common.BinaryMessenger
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import java.nio.ByteBuffer
import java.util.ArrayList
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test

class MultipleArityTests {

  @Test
  fun testSimpleHost() {
    val binaryMessenger = mockk<BinaryMessenger>()
    val api = mockk<MultipleArityHostApi>()

    val inputX = 10L
    val inputY = 5L

    val channelName = "dev.flutter.pigeon.pigeon_integration_tests.MultipleArityHostApi.subtract"
    val handlerSlot = slot<BinaryMessenger.BinaryMessageHandler>()

    every { binaryMessenger.setMessageHandler(channelName, capture(handlerSlot)) } returns Unit
    every { api.subtract(any(), any()) } answers { firstArg<Long>() - secondArg<Long>() }

    MultipleArityHostApi.setUp(binaryMessenger, api)

    val codec = MultipleArityHostApi.codec
    val message = codec.encodeMessage(listOf(inputX, inputY))
    message?.rewind()
    handlerSlot.captured.onMessage(message) {
      it?.rewind()
      @Suppress("UNCHECKED_CAST") val wrapped = codec.decodeMessage(it) as List<Any>?
      assertNotNull(wrapped)
      wrapped?.let { assertEquals(inputX - inputY, wrapped[0]) }
    }
  }

  @Test
  fun testSimpleFlutter() {
    val binaryMessenger = mockk<BinaryMessenger>()
    val api = MultipleArityFlutterApi(binaryMessenger)

    val inputX = 10L
    val inputY = 5L

    every { binaryMessenger.send(any(), any(), any()) } answers
        {
          val codec = MultipleArityFlutterApi.codec
          val message = arg<ByteBuffer>(1)
          val reply = arg<BinaryMessenger.BinaryReply>(2)
          message.position(0)
          val args = codec.decodeMessage(message) as ArrayList<*>
          val argX = args[0] as Long
          val argY = args[1] as Long
          val replyData = codec.encodeMessage(listOf(argX - argY))
          replyData?.position(0)
          reply.reply(replyData)
        }

    var didCall = false
    api.subtract(inputX, inputY) {
      didCall = true
      assertEquals(inputX - inputY, it.getOrNull())
    }

    assertTrue(didCall)
  }
}
